### ISSUE TITLE
拿不到wasm的返回值，所有的方法的返回值都是undefined

### ISSUE THREAD

**Issue by caicai1998** [Menci/vite-plugin-wasm#1]

**Comment by Menci** [comment]
可否给一个例子？

以及，你的代码在 webpack5 可以正常工作吗

**Comment by caicai1998** [comment]
> 可否给一个例子？
> 
> 以及，你的代码在 webpack5 可以正常工作吗

wasm编辑发布在npm上，webpack5正常运行。在vite底下执行了wasm方法，但是没有返回值

**Comment by Menci** [comment]
可否提供一个最小的可以复现的项目

根据您目前提供的描述，我无法了解到任何有用的信息

**Comment by Menci** [comment]
ping @caicai1998

**Comment by ShravanSunder** [comment]
@Menci i'm having the same issue.  I'm importing @ironcorelabs/recrypt-wasm-binding;

my vite config has 
```

export default defineConfig({
  plugins: [
    wasm(),
```

i've also tried to exclude`optimizeDeps` for that npm package.

**Comment by Menci** [comment]
@sigoden Do you have any ideas...

**Comment by sigoden** [comment]
To make wasm work

1. vite >= 3.1.2
2.  exclude wasm module from pre-bundling
```
export default defineConfig({
  optimizeDeps: {
    exclude: [
      "<wasm node module>"
    ]
  },
  plugins: [
    wasm(),
    topLevelAwait()
  ]
})

```

**Comment by ShravanSunder** [comment]
Yes that's essentially what i've done.  It works in nextjs app.  but in vite with this plugin, i'm having issues. cc: @Menci @sigoden 

The functions all return undefined.

Here is the full config, its a bit big.   The package that's being excluded is `'@ironcorelabs/recrypt-wasm-binding'`

```typescript
import { resolve } from 'path';

import reactPlugin from '@vitejs/plugin-react';
import chalk from 'chalk';
import { defineConfig } from 'vite';
import checker from 'vite-plugin-checker';
import { viteExternalsPlugin } from 'vite-plugin-externals';
import wasm from 'vite-plugin-wasm';
import tsconfigPaths from 'vite-tsconfig-paths';

const isDev = process.env.ENVIRONMENT === 'DEVELOPMENT';
console.log('env.dev:', process.env.ENVIRONMENT, ' isDev:', isDev);
console.log();
console.log(
  chalk.green('🏁 Make sure to build the contracts with `yarn contracts:build` and deploy them with `yarn deploy`')
);
console.log();

/**
 * These libraries should not be egarly bundled by vite.  They have strange dependencies and are not needed for the app.
 */
const excludeDeps = ['@apollo/client', `graphql`, 'electron', 'electron-fetch', '@ironcorelabs/recrypt-wasm-binding'];

/**
 * browserify for web3 components
 */
const externals = {
  http: resolve(__dirname, './node_modules/http-browserify'),
  https: resolve(__dirname, './node_modules/http-browserify'),
  timers: resolve(__dirname, './node_modules/timers-browserify'),
  // the two below are due to bad ipfs-core dependency, they are not loaded
  electron: 'electron',
  'electron-fetch': 'electron-fetch',
};

const nodeShims = {
  util: resolve(__dirname, './node_modules/util'),
};

/**
 * Externals:
 * - node externals are required because web3 are terribly bundled and some of them use commonjs libraries.  modern libs like ethers help with this.
 * - electron:  added due to ipfs-http-client.  it has very poor esm compatibility and a ton of dependency bugs. see: https://github.com/ipfs/js-ipfs/issues/3452
 */
const externalPlugin = viteExternalsPlugin({
  ...externals,
  ...(isDev ? { ...nodeShims } : {}),
});

export default defineConfig({
  plugins: [
    wasm(),
    reactPlugin({
      jsxRuntime: 'automatic',
      jsxImportSource: `@emotion/react`,
    }),
    tsconfigPaths(),
    externalPlugin,
    checker({ typescript: true }),
  ],
  build: {
    minify: false,
    sourcemap: true,
    commonjsOptions: {
      include: /node_modules/,
      transformMixedEsModules: true,
    },
    rollupOptions: {
      input: {
        main: resolve(__dirname, 'index.html'),
      },
    },
  },
  define: {
    'process.env': {},
  },
  optimizeDeps: {
    esbuildOptions: {
      target: 'es2020',
    },
    exclude: excludeDeps,
    include: ['@scaffold-eth/common', 'eth-hooks', 'eth-components', '@ironcorelabs/recrypt-wasm-binding'],
  },
  json: {
    namedExports: true,
  },
  resolve: {
    preserveSymlinks: true,
    mainFields: ['module', 'main', 'browser'],
    alias: {
      '~~': resolve(__dirname, 'src'),
      '~common': resolve(__dirname, '../common/src'),
      // -------------------------------------------
      // your aliases
      'eth-hooks': resolve(__dirname, './node_modules/eth-hooks'),
      'eth-components': resolve(__dirname, './node_modules/eth-components'),
      'react-css-theme-switcher': resolve(__dirname, './node_modules/react-css-theme-switcher'),
      react: resolve(__dirname, './node_modules/react'),
      'react-dom': resolve(__dirname, './node_modules/react-dom'),
      // -------------------------------------------
      ...externals,
      ...nodeShims,
      // required by web3 dependencies
      process: 'process',
      stream: 'stream-browserify',
    },
  },
  server: {
    hmr: true,
    port: 3000,
    watch: {
      followSymlinks: true,
    },
    fs: {
      // compatability for yarn workspaces
      allow: ['../../'],
    },
  },
  css: {
    preprocessorOptions: {
      less: {
        javascriptEnabled: true,
      },
    },
  },
});

```

**Comment by sigoden** [comment]
![image](https://user-images.githubusercontent.com/4012553/192680044-e5c23a57-2b75-44b3-9bb1-3e0157272605.png)

you include it after exclusion. @ShravanSunder

**Comment by ShravanSunder** [comment]
@sigoden doh! 🤦🏽‍♂️ i must have made mistake in trying to make it work.

thank you!

**Comment by Menci** [comment]
I believe this issue has been fixed since `v3.1.0` even if building with `optimizeDeps`. Closing it now.
